File: backend/src/entities/Article.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn,
  OneToMany,
} from "typeorm";
import { User } from "./User";
import { ArticleCategory } from "./ArticleCategory";
import { ArticleTag } from "./ArticleTag";

export enum ArticleStatus {
  DRAFT = "draft",
  SUBMITTED = "submitted",
  UNDER_REVIEW = "under_review",
  APPROVED = "approved",
  REJECTED = "rejected",
  PUBLISHED = "published",
}

@Entity("articles")
export class Article {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column()
  title!: string;

  @Column({ nullable: true })
  subtitle?: string;

  @Column({ unique: true })
  slug!: string;

  @Column("text", { nullable: true })
  excerpt?: string;

  @Column("text")
  contentJson!: string; // TipTap JSONContent as string

  @Column("text", { nullable: true })
  contentHtml?: string; // Computed HTML from JSON

  @Column({
    type: "varchar",
    default: ArticleStatus.DRAFT,
  })
  status!: ArticleStatus;

  @Column()
  authorId!: number;

  @ManyToOne(() => User, (user) => user.articles)
  @JoinColumn({ name: "authorId" })
  author!: User;

  @Column({ nullable: true })
  featuredImage?: string;

  @Column({ nullable: true })
  seoTitle?: string;

  @Column("text", { nullable: true })
  seoDescription?: string;

  @Column("text", { nullable: true })
  seoKeywords?: string;

  @Column({ default: false })
  isFeatured!: boolean;

  @Column({ default: false })
  isBreakingNews!: boolean;

  @Column({ default: true })
  allowComments!: boolean;

  @Column({ default: "web,mobile" })
  visibility!: string;

  @Column({ nullable: true })
  scheduledAt?: Date;

  @Column({ nullable: true })
  publishedAt?: Date;

  @Column("text", { nullable: true })
  reviewNotes?: string;

  @Column({ default: 0 })
  wordCount!: number;

  @Column({ default: 0 })
  characterCount!: number;

  @Column({ default: 0 })
  readingTime!: number;

  @Column({ default: 0 })
  views!: number;

  @OneToMany(
    () => ArticleCategory,
    (articleCategory) => articleCategory.article
  )
  articleCategories!: ArticleCategory[];

  @OneToMany(() => ArticleTag, (articleTag) => articleTag.article)
  articleTags!: ArticleTag[];

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;
}

File: backend-dotnet/Vietsov.Api/Models/Article.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Vietsov.Api.Models;

public enum ArticleStatus
{
    Draft,
    Submitted,
    UnderReview,
    Approved,
    Rejected,
    Published
}

public class Article
{
    [Key]
    public int Id { get; set; }

    [Required]
    [MaxLength(500)]
    public string Title { get; set; } = string.Empty;

    [MaxLength(500)]
    public string? Subtitle { get; set; }

    [Required]
    [MaxLength(500)]
    public string Slug { get; set; } = string.Empty;

    [Column(TypeName = "text")]
    public string? Excerpt { get; set; }

    [Required]
    [Column(TypeName = "text")]
    public string ContentJson { get; set; } = string.Empty; // TipTap JSONContent as string

    [Column(TypeName = "text")]
    public string? ContentHtml { get; set; } // Computed HTML from JSON

    [Required]
    public ArticleStatus Status { get; set; } = ArticleStatus.Draft;

    [Required]
    public int AuthorId { get; set; }

    [ForeignKey("AuthorId")]
    public virtual ApplicationUser Author { get; set; } = null!;

    [MaxLength(500)]
    public string? FeaturedImage { get; set; }

    [MaxLength(500)]
    public string? SeoTitle { get; set; }

    [Column(TypeName = "text")]
    public string? SeoDescription { get; set; }

    [Column(TypeName = "text")]
    public string? SeoKeywords { get; set; }

    public bool IsFeatured { get; set; } = false;

    public bool IsBreakingNews { get; set; } = false;

    public bool AllowComments { get; set; } = true;

    [MaxLength(100)]
    public string Visibility { get; set; } = "web,mobile";

    public DateTime? ScheduledAt { get; set; }

    public DateTime? PublishedAt { get; set; }

    [Column(TypeName = "text")]
    public string? ReviewNotes { get; set; }

    public int WordCount { get; set; } = 0;

    public int CharacterCount { get; set; } = 0;

    public int ReadingTime { get; set; } = 0;

    public int Views { get; set; } = 0;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    // Navigation properties
    public virtual ICollection<ArticleCategory> ArticleCategories { get; set; } = new List<ArticleCategory>();
    public virtual ICollection<ArticleTag> ArticleTags { get; set; } = new List<ArticleTag>();
}

